scripts/nn/layers/graph_conv.dml

```diff
@@ -0,0 +1,262 @@
+             w.r.t. the input feature matrix X, matrix of shape (n, f_in).
+     * - dW: gradient of the linear weights i.e. the partial derivative of the loss w.r.t.
+     *       the linear weights W, matrix of shape (f_in, f_out).
+     * - db: gradient of the bias i.e. the partial derivative of the loss w.r.t. the bias b,
+     *       matrix of shape (1, f_out).
+     */
+    n = nrow(X)
+    m = nrow(edge_index)
+
+    # count degrees for normalization
+    if (add_self_loops)
+    {
+        # add self loops and count degrees
+        D_in = matrix(1, n, 1)
+
+        loop_index = seq(1,n,1) %*% matrix(1, rows=1, cols=2)
+        loop_weight = matrix(1, rows=n, cols=1)
+
+        for (j in 1:m)
+        {
+            # count degree according to the weight for correct normalization
+            D_in[as.integer(as.scalar(edge_index[j, 2])), 1] += edge_weight[j, 1]
+
+            # for self loops
+            if (as.scalar(edge_index[j, 1]) == as.scalar(edge_index[j, 2]))
+            {
+                if (as.scalar(loop_weight[as.integer(as.scalar(edge_index[j, 1]))]) != 0.0)
+                {
+                    loop_weight[as.integer(as.scalar(edge_index[j, 1]))] = 0
+                    # remove 1 degree again that was added by initializing D_in with 1
+                    D_in[as.integer(as.scalar(edge_index[j, 2])), 1] += -1.0
+                }
+            }
+        }
+
+        edge_index = rbind(edge_index, loop_index)
+        edge_weight = rbind(edge_weight, loop_weight)
+    }
+    else
+    {
+        # count degrees
+        D_in = matrix(0, n, 1)
+        for (j in 1:m)
+        {
+            # count degree according to the weight for correct normalization
+            node_in = as.integer(as.scalar(edge_index[j, 2]))
+            D_in[node_in, 1] += edge_weight[j, 1]
+        }
+    }
+
+    # message passing/convolve dOut with reversed edges to compute A_hat^T * dOut
+    dOut_agg_rev = matrix(0, rows=nrow(dOut), cols=ncol(dOut))
+    m = nrow(edge_index)
+    for (j in 1:m)
+    {
+        # doing: edge_weight[j] *= 1/sqrt(degree(node_in)*degree(node_out))
+        if (as.scalar(D_in[as.integer(as.scalar(edge_index[j, 2])), 1] *
+            D_in[as.integer(as.scalar(edge_index[j, 1])), 1]) != 0)
+        {
+            edge_weight[j, 1] = edge_weight[j, 1] * (1 / sqrt(D_in[as.integer(as.scalar(edge_index[j, 2])), 1]*
+                                                              D_in[as.integer(as.scalar(edge_index[j, 1])), 1]))
+        }
+        else
+        {
+            edge_weight[j, 1] = 0.0
+        }
+        dOut_agg_rev[as.integer(as.scalar(edge_index[j, 1]))] += as.scalar(edge_weight[j, 1]) * dOut[as.integer(as.scalar(edge_index[j, 2]))]
+    }
+
+    # calculate gradient w.r.t. X (Formula: A_hat^T * dOut * W^T)
+    dX = dOut_agg_rev %*% t(W)
+
+    # calculate gradient w.r.t. W (Formula: X^T * A_hat^T * dOut)
+    dW = t(X) %*% dOut_agg_rev
```

Comment:
or is it the gradient taking time?

Reply:
In the stress test, I only used the forward pass.